Update RTC-S1 SAS for BETA delivery

gshiroma opened the pull request:
This PR brings the following updates over the previous [Interface (IF) Release v0.1](https://github.com/opera-adt/RTC/releases/tag/v0.1):

Add option to save RTC-S1 imagery as GeoTIFFs and metadata as an HDF5.

seongsujeong commented on app/rtc_s1.py:
```diff
@@ -475,80 +669,163 @@ def run(cfg):
+                     f'{imagery_extension}')
+                output_burst_imagery_list.append(geo_burst_pol_filename)
+
+            _separate_pol_channels(geo_burst_filename,
```
'_separate_pol_channels()` will save the image into geotiff format regardless of `imagery_extension`.

seongsujeong commented on src/rtc/core.py:
```diff
@@ -1,7 +1,11 @@
 import sys
+import shutil
 import logging
+import tempfile
```
nit - sort the import order above

gshiroma replied:
I usually import the most generic library first, which sometimes can be subjective. Stardard libraries before third-party libraries and both before local libraries. Do you have any other suggestions? You are probably thinking of using the alphabetical order?

seongsujeong replied:
Yes, I meant the alphabetical order for the standard libraries, but let's keep the order as it is as we discussed.

seongsujeong commented on src/rtc/core.py:
```diff
@@ -47,6 +51,102 @@ def flush(self):
+        logger = logging.getLogger('proteus')
+
+    logger.info('        COG step 1: add overviews')
+    gdal_ds = gdal.Open(filename, 1)
```
```suggestion
    gdal_ds = gdal.Open(filename, gdal.GA_Update)
```
For clarity

seongsujeong commented on src/rtc/core.py:
```diff
@@ -47,6 +51,102 @@ def flush(self):
+    if compression:
+        gdal_translate_options += [f'COMPRESS={compression}']
+
+    is_integer = 'byte' in dtype_name  or 'int' in dtype_name
```
Looks like the same code as around line 88

seongsujeong commented on src/rtc/geogrid.py:
```diff
@@ -266,7 +266,7 @@ def get_point_epsg(lat, lon):
 def generate_geogrids(bursts, geo_dict, dem):
 
-    dem_raster = isce3.io.Raster(dem)
+    # dem_raster = isce3.io.Raster(dem)
```
Please add the doctoring for `generate_geogrids()`

seongsujeong commented on src/rtc/h5_prep.py:
```diff
@@ -141,7 +182,9 @@ def populate_metadata_group(h5py_obj: h5py.File,
         #   [int(burst_in.burst_id[1:4]), 'Relative orbit number'],
         'identification/trackNumber':
-            [int(burst_in.burst_id.split('_')[1]), 'Track number'],
+            [int(str(burst_in.burst_id).split('_')[1]), 'Track number'],
```
```suggestion
            [burst_in.burst_id.esa_burst_id, 'Track number'],
```

seongsujeong commented on src/rtc/mosaic_geobursts.py:
```diff
@@ -238,27 +238,133 @@ def weighted_mosaic(list_rtc_images, list_nlooks, geo_filename,
+
+    '''
+    mosaic_dict = _weighted_mosaic(list_rtc_images, list_nlooks,
+                                     geogrid_in=geogrid_in, verbose = verbose)
```
```suggestion
                                   geogrid_in=geogrid_in, verbose = verbose)
```

seongsujeong replied:
indentation

seongsujeong commented on src/rtc/mosaic_geobursts.py:
```diff
@@ -103,22 +103,22 @@ def check_reprojection(geogrid_mosaic,
+                     geogrid_in=None, verbose = True):
     '''
-    Mosaic the snapped S1 geobursts
+    Mosaic S-1 geobursts and return mosaic dictionary
```
```suggestion
    Mosaic S-1 geobursts and return the mosaic as dictionary
```

seongsujeong commented on src/rtc/mosaic_geobursts.py:
```diff
@@ -103,22 +103,22 @@ def check_reprojection(geogrid_mosaic,
 
-def weighted_mosaic(list_rtc_images, list_nlooks, geo_filename,
-                    geogrid_in=None, verbose = True):
+def _weighted_mosaic(list_rtc_images, list_nlooks,
```
The function name is a bit confusing because we have `weighted_mosaic()` in the following function definition. How about something like `_compute_weighted_mosaic()`?

seongsujeong reviewed (commented):
Thanks @gshiroma for the PR. Here I leave some comments.

I see dome PEP8 issues throughout the code. I think we should clean them up as much as we can before the beta delivery.
Once this PR is merged, I can run the lint, fix the issues, and submit the PR for that.

gshiroma commented:
> 

Thank you, @seongsujeong , for your review! I have just one comment left to address regarding the import order. I think it's a great idea to use some lint tool after this PR is approved/merged.

seongsujeong reviewed (approved):
Looks good to me. Thanks for the quick turnaround!

gshiroma commented:
Thank you, @seongsujeong !

gshiroma commented:
@seongsujeong , thank you for your recent updates! Just wondering if you'd like to take another look at this PR to make sure it can be merged.
@hfattahi , this PR should be ready for review.

seongsujeong commented on src/rtc/h5_prep.py:
```diff
@@ -141,7 +188,9 @@ def populate_metadata_group(h5py_obj: h5py.File,
-            [int(burst_in.burst_id.split('_')[1]), 'Track number'],
+            [burst_in.burst_id.esa_burst_id, 'Track number'],
+        'identification/boundingPolygon':
+            [get_polygon_wkt(burst_in), 'bounding polygon of Burst as WKT'],
```
```suggestion
            [get_polygon_wkt(burst_in), 'bounding polygon of Burst in WKT'],
```

seongsujeong commented on src/rtc/h5_prep.py:
```diff
@@ -141,7 +188,9 @@ def populate_metadata_group(h5py_obj: h5py.File,
+            [get_polygon_wkt(burst_in), 'bounding polygon of Burst as WKT'],
         'identification/missionId':
             [burst_in.platform_id, 'Mission identifier'],
         # NOTE maybe `SLC` has to be sth. like RTC?
```
```suggestion

```

seongsujeong reviewed (approved):
LGTM. Just minor fixes as in the comments below :)

hfattahi commented on src/rtc/defaults/rtc_s1.yaml:
```diff
@@ -1,147 +1,183 @@
+            # Required. List of orbit (EOF) files (min=1)
+            orbit_file_path:
+            # Required. The unique burst ID to process
+            burst_id:
```
Is burst_id really required? I think as an optional parameter it is fine to have. But for operation, we want SDS to take one Safe file and generate all burst products. Basically the default should be to have this field empty and process all bursts.

gshiroma replied:
Oh, good catch . `burst_id` is not required in the code. I'm going to fix the comment.

hfattahi commented on src/rtc/defaults/rtc_s1.yaml:
```diff
@@ -1,147 +1,183 @@
+            dem_description:
+
+            # burst database sqlite file
+            burst_database_file:
```
Thanks for adding the database. However the database is supposed to be static ancillary file and not dynamic.

gshiroma replied:
Got it, will update the runconfig and code accordingly. Thanks.

hfattahi reviewed (commented):
Very impressive and a lot of additions to the workflow. I have very minor comments before I approve.

gshiroma commented:
@hfattahi and @seongsujeong , I had to push a few last-minutes commits to this PR.  I changed a few defaults for the runconfig file that we expect to be used as a reference by the SDS, and I decoupled the location where RTC-S1 imagery and secondary layers are saved. Before these PRs, if the RTC-S1 imagery were saved outside of the HDF5 file, the secondary layers (e.g., layover shadow mask) would also be saved outside of the HDF5 file. Now, I added a flag `save_secondary_layers_as_hdf5` that provides an option to save the secondary layers inside or outside the HDF5 file regardless of the RTC-S1 imagery.

hfattahi commented:
Thanks @gshiroma . I was able to successfully run the workflow on a test case. Would you please add "burst_id" to the "/science/SENTINEL1/identification"?

hfattahi commented:
@gshiroma I'm looking at the orbit stored in the product. I see that the orbit time are stored as seconds from a reference epoch. However, the ref epoch is not stored in the product. Would you please either change the time to "UTC" or add the ref epoch to the same orbit group?

hfattahi commented:
Probably beyond this PR but I'm just wondering what is our current plan for full transparency on the reproducibility of the product. I like that we have a clear list of input files (S1 safe file and orbit). However all parameters of the run config is not stored in the processing information. I think either SAS or SDS needs to keep track of these parameters. That way if we want to reproduce the product, we would cerate the run config, download the inputs and would use the correct software version. @gshiroma @seongsujeong any opinion on this?

hfattahi commented:
Great to see the isce version in ```'/science/SENTINEL1/RTC/metadata/processingInformation/algorithms'``` . but I think we should keep track of the ```RTC version``` and ```s1-reader version``` as well. 
Also remember that we had a long discussion about the product versioning. Please make sure that the ```product_version``` is a field in the run config which SDS will populate with the correct value, the SAS will read it and writes it to the product metadata.

seongsujeong commented:
@hfattahi Good point. MAybe we need to add those parameters when writing the metadata.
There has been some addition of the parameters since the interface release. For example, we now have burst DB file, and rdr2geo parameters for calculating layover shadow mask.

@gshiroma  Do you think this needs to happen in the beta release? I think about 0.5 day would be enough to add them as soon as we identify what to add.

hfattahi commented:
In the test I did, I see that the shadow/layover is created and stored in the metadata hdf5 file. Just double checking if we made a decision on this or it is still floating? Basically will every product have the shadow layover mask?

m[f'/science/SENTINEL1/RTC/grids/frequencyA/layoverShadowMask']

seongsujeong commented:
> Basically will every product have the shadow layover mask?
We need more study on this to make decision if we need to calculate layover shadow masks every time. What I've seen so far was some changes in layover shadow mask over time (Colorado, 03/2022 - 09/2022).
The changes could be due to baseline (changes up to ~150m), but could be also because of the resampling method we use for layover shadow mask (i.e. nearest neighbor).

gshiroma commented:
> Thanks @gshiroma . I was able to successfully run the workflow on a test case. Would you please add "burst_id" to the "/science/SENTINEL1/identification"?
> 
> what are you planning to put in "processingType" which is currently "b'UNDEFINED'" ?
> 
> track number is populated incorrectly as "trackNumber 371795"

@hfattahi , I added `burst_id` to `/science/SENTINEL1/identification` and `processing_type` to the runconfig.

I don't understand why the track number is incorrect. How should it look like?

gshiroma commented:
> @gshiroma I'm looking at the orbit stored in the product. I see that the orbit time are stored as seconds from a reference epoch. However, the ref epoch is not stored in the product. Would you please either change the time to "UTC" or add the ref epoch to the same orbit group?

@hfattahi , I added `referenceEpoch` to orbit. Please, let me know if that looks OK now. We  should replicate it on the NISAR products.

hfattahi commented:
> > Thanks @gshiroma . I was able to successfully run the workflow on a test case. Would you please add "burst_id" to the "/science/SENTINEL1/identification"?
> > what are you planning to put in "processingType" which is currently "b'UNDEFINED'" ?
> > track number is populated incorrectly as "trackNumber 371795"
> 
> @hfattahi , I added `burst_id` to `/science/SENTINEL1/identification` and `processing_type` to the runconfig.
> 
> I don't understand why the track number is incorrect. How should it look like?

I mean we have maximum 175 tracks. The value in the product is 371795.

gshiroma commented:
> > > Thanks @gshiroma . I was able to successfully run the workflow on a test case. Would you please add "burst_id" to the "/science/SENTINEL1/identification"?
> > > what are you planning to put in "processingType" which is currently "b'UNDEFINED'" ?
> > > track number is populated incorrectly as "trackNumber 371795"
> > 
> > 
> > @hfattahi , I added `burst_id` to `/science/SENTINEL1/identification` and `processing_type` to the runconfig.
> > I don't understand why the track number is incorrect. How should it look like?
> 
> I mean we have maximum 175 tracks. The value in the product is 371795.

Thank you @hfattahi . @seongsujeong , you may want to revise that piece of code.

gshiroma commented:
> Great to see the isce version in `'/science/SENTINEL1/RTC/metadata/processingInformation/algorithms'` . but I think we should keep track of the `RTC version` and `s1-reader version` as well. Also remember that we had a long discussion about the product versioning. Please make sure that the `product_version` is a field in the run config which SDS will populate with the correct value, the SAS will read it and writes it to the product metadata.

I added the `product_version` to the runconfig and now we are using it to populate the `productVersion` in the HDF5 metadata

gshiroma commented:
> > > Thanks @gshiroma . I was able to successfully run the workflow on a test case. Would you please add "burst_id" to the "/science/SENTINEL1/identification"?
> > > what are you planning to put in "processingType" which is currently "b'UNDEFINED'" ?
> > > track number is populated incorrectly as "trackNumber 371795"
> > 
> > 
> > @hfattahi , I added `burst_id` to `/science/SENTINEL1/identification` and `processing_type` to the runconfig.
> > I don't understand why the track number is incorrect. How should it look like?
> 
> I mean we have maximum 175 tracks. The value in the product is 371795.

Yes, @hfattahi , the layover/shadow mask is the only secondary layer that we are thinking about keeping in the product. But this is still [TBC]

gshiroma commented:
Hi @hfattahi , @seongsujeong fixed the track number and updated the Dockerfile for this BETA delivery. Now, we are just waiting for your approval to merge and tag this PR.

hfattahi reviewed (approved):
Thanks for addressing my comments. Looks good to me.
